File: editors/vscode/src/lspext.ts
import * as lc from "vscode-languageclient";

export interface SyntaxTreeParams {
  textDocument: lc.TextDocumentIdentifier;
  range?: lc.Range;
}

export const syntaxTree = new lc.RequestType<SyntaxTreeParams, string, void>(
  "pdls/syntaxTree"
);

File: src/pd-ls/src/lsp_ext.rs
use lsp_types::request::Request;
use lsp_types::{Range, TextDocumentIdentifier};
use serde::{Deserialize, Serialize};

pub enum SyntaxTree {}

impl Request for SyntaxTree {
    type Params = SyntaxTreeParams;
    type Result = String;

    const METHOD: &'static str = "pdls/syntaxTree";
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct SyntaxTreeParams {
    pub text_document: TextDocumentIdentifier,
    pub range: Option<Range>,
}
